tests/runtime/test_http_server.py

```diff
@@ -0,0 +1,255 @@
+
+
+@dataclass
+class SampleServer:
+    server: http_server.RuntimeHTTPServer
+    port: int
+    ca_certfile: Optional[str]
+    client_keyfile: Optional[str]
+    client_certfile: Optional[str]
+
+
+@dataclass
+class KeyPair:
+    cert: str
+    key: str
+
+
+@dataclass
+class TLSConfig:
+    server: KeyPair
+    client: KeyPair
+
+
+@contextmanager
+def generate_tls_configs(
+    tls: bool = False, mtls: bool = False, **http_config_overrides
+):
+    """Helper to generate tls configs"""
+    with tempfile.TemporaryDirectory() as workdir:
+        config_overrides = {}
+        client_keyfile, client_certfile, ca_certfile = None, None, None
+        if mtls or tls:
+            ca_key = tls_test_tools.generate_key()[0]
+            ca_cert = tls_test_tools.generate_ca_cert(ca_key)
+            ca_certfile, _ = save_key_cert_pair("ca", workdir, cert=ca_cert)
+            server_certfile, server_keyfile = save_key_cert_pair(
+                "server",
+                workdir,
+                *tls_test_tools.generate_derived_key_cert_pair(ca_key=ca_key),
+            )
+
+            tls_config = TLSConfig(
+                server=KeyPair(cert=server_certfile, key=server_keyfile),
+                client=KeyPair(cert="", key=""),
+            )
+            # need to save this ca_certfile in config_overrides so the tls tests below can access it from client side
+            config_overrides["use_in_test"] = {"ca_cert": ca_certfile}
+            if mtls:
+                client_certfile, client_keyfile = save_key_cert_pair(
+                    "client",
+                    workdir,
+                    *tls_test_tools.generate_derived_key_cert_pair(ca_key=ca_key),
+                )
+                tls_config.client = KeyPair(cert=ca_certfile, key="")
+                # need to save the client cert and key in config_overrides so the mtls test below can access it
+                config_overrides["use_in_test"]["client_cert"] = client_certfile
+                config_overrides["use_in_test"]["client_key"] = client_keyfile
+
+            config_overrides["runtime"] = {"tls": tls_config}
+        port = http_server.RuntimeServerBase._find_port()
+        config_overrides.setdefault("runtime", {})["http"] = {
+            "timeout": 0.1,  # this is so the server is killed after 0.1 if no test is running
+            "port": port,
+            **http_config_overrides,
+        }
+
+        with temp_config(config_overrides, "merge"):
+            yield config_overrides
+
+
+## Insecure and TLS Tests #######################################################################
+
+
+def test_insecure_server():
+    with generate_tls_configs():
+        insecure_http_server = http_server.RuntimeHTTPServer()
+        with insecure_http_server.run_in_thread():
+            resp = requests.get(f"http://localhost:{insecure_http_server.port}/docs")
+            resp.raise_for_status()
+
+        insecure_http_server._stop_metering()
```

Comment:
We should get this into `def stop` so that it can be in the context manager automatically